core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -1038,12 +1052,14 @@ private object RTS {
     }
 
     private final def purgeJoinersKillers(v: ExitResult[E, A],
+                                          exitHandler: Option[ExitResult[E, A] => IO[Nothing, Unit]],
                                           joiners: List[Callback[E, A]],
                                           killers: List[Callback[E, Unit]]): Unit = {
       // To preserve fair scheduling, we submit all resumptions on the thread
       // pool in (rough) order of their submission.
       killers.reverse.foreach(k => rts.submit(k(SuccessUnit)))
       joiners.foreach(k => rts.submit(k(v)))
+      exitHandler.foreach(k => rts.submit(evaluate(k(v))))
```

Comment:
We cannot call `evaluate` here because `evaluate` belongs to this fiber, with its own stack, etc. You could call fork to run it on a new fiber (maybe we need a helper to do this?) or `unsafeRunAsync`, perhaps (since the result value is not needed).

Reply:
I think `unsafeRunAsync` does not fully work as we still need to wait for the termination of the exit handlers, don’t we? As a new fiber is created, it escapes the control of the root fiber, doesn’t it? Which would explain why my test is failing.

Reply:
It's true it escapes the control of the root fiber, but it's an asynchronous callback and therefore kind of expected (I think). If you show me which test is failing I can try to see if it's legit failure or not.